apps/hash-frontend/src/pages/[shortname].page/edit-pinned-entity-types-modal.tsx

```diff
@@ -0,0 +1,424 @@
+
+  useEffect(() => {
+    element.style.pointerEvents = "none";
+    element.style.position = "absolute";
+    element.style.height = "100%";
+    element.style.width = "100%";
+    element.style.top = "0";
+
+    document.body.appendChild(element);
+
+    return () => {
+      document.body.removeChild(element);
+    };
+  }, [element]);
+
+  return (render: (provided: DraggableProvided) => ReactElement) =>
+    (provided: DraggableProvided) => {
+      const result = render(provided);
+
+      const style = provided.draggableProps.style as
+        | DraggingStyle
+        | NotDraggingStyle;
+
+      if ("position" in style) {
+        return createPortal(result, element);
+      }
+      return result;
+    };
+};
+
+type PinnedEntityTypesFormData = {
+  pinnedEntityTypes: EntityTypeWithMetadata[];
+};
+
+export const EditPinnedEntityTypesModal: FunctionComponent<
+  Omit<ModalProps, "children" | "onClose"> & {
+    onClose: () => void;
+    profile: User | Org;
+    refetchProfile: () => Promise<void>;
+  }
+> = ({ profile, onClose, refetchProfile, ...modalProps }) => {
+  const { authenticatedUser } = useAuthenticatedUser();
+  const { entityTypes, isSpecialEntityTypeLookup } =
+    useEntityTypesContextRequired();
+
+  const [loading, setLoading] = useState(false);
+  const [displayEntityTypesSearch, setDisplayEntityTypesSearch] =
+    useState(false);
+
+  const { updateEntity } = useBlockProtocolUpdateEntity();
+  const { createEntityType } = useBlockProtocolCreateEntityType(
+    authenticatedUser.accountId as OwnedById,
+  );
+
+  const { control, reset, handleSubmit, formState } =
+    useForm<PinnedEntityTypesFormData>({
+      defaultValues: {
+        pinnedEntityTypes: [],
+      },
+    });
+
+  const initializedFields = useRef(false);
+
+  useEffect(() => {
+    if (entityTypes && !initializedFields.current) {
+      const initialPinnedEntityTypes = entityTypes.filter(
+        ({ metadata }) =>
+          profile.pinnedEntityTypeBaseUrls?.includes(metadata.recordId.baseUrl),
+      );
+
+      initializedFields.current = true;
+      reset({ pinnedEntityTypes: initialPinnedEntityTypes });
+    }
+  }, [entityTypes, profile.pinnedEntityTypeBaseUrls, reset]);
+
+  const { fields, append, remove, replace } = useFieldArray({
+    control,
+    name: "pinnedEntityTypes",
+  });
+
+  const innerSubmit = handleSubmit(async (data) => {
+    setLoading(true);
+
+    const { pinnedEntityTypes } = data;
+
+    await updateEntity({
+      data: {
+        entityId: profile.entity.metadata.recordId.entityId,
+        entityTypeId: profile.entity.metadata.entityTypeId,
+        properties: {
+          ...profile.entity.properties,
+          [extractBaseUrl(
+            types.propertyType.pinnedEntityTypeBaseUrl.propertyTypeId,
+          )]: pinnedEntityTypes.map(
+            ({ metadata }) => metadata.recordId.baseUrl,
+          ),
+        },
+      },
+    });
+
+    /** @todo: error handling */
+
+    setLoading(false);
```

Comment:
If you called `updateEntity` via `useMutation` rather than the BP hook you could use the loading and error return from the ApolloClient hook rather than manually here. Maybe not worth bothering with since we will remove ApolloClient at some point, but probably not for a while and we will replace it with some similar hook when we do

Reply:
Good idea, have done so in https://github.com/hashintel/hash/pull/3338/commits/ebe7262785fe3b48e66d9e4966b52014a2393208